### DISCUSSION TITLE
My late night project for openmage v21

### DISCUSSION THREAD

**Discussion by fballiano** [OpenMage/magento-lts#2270] [Show & Tell]
So, you all know I want to remove a lot of code from OM's repo and delegate responsability to the original authors, I a forked OM v20 on my repo and started the work:
https://github.com/fballiano/openmage/tree/mcommerce

I called it mcommerce at the moment, I don't really know why but that's beside the point.

Notes:
- based on OM 20
- bumped requirement to >=php7.4
- moved to zf1-future installed via composer
- moved net_idna2, linuxlibertine-fonts, phpseclib, mcrypt_compat to composer requirements
- moved colinmollenhour/cache-backend-redis and colinmollenhour/magento-redis-session to composer suggest
- added some other modules to composer suggest (like image cleaner and full page cache)
- updated emogrifier to 6.0 (installed via composer)
- removed web based installation system (you can install via command line or magerun)

Preliminary tests are ok and I'm really liking it. I'd obviously a little "edgy" but I think that is the direction to go.

I'd like it to be the next OM and I hope it will not take years to bring some more news into OM.

Let me know what you think!

**Comment by kiatng** [comment]
Next OM, modernize the js framework for use in both front and backend.

**Reply by fballiano** [reply]
all of the current changes are compatible with the past but what scares me about the javascript part is that most of the frontend modules (and payment methods) won't work anymore

**Comment by addison74** [comment]
It is difficult to express an opinion at this moment, but I have gone through such situations when certain projects fell apart due to contrary ideas, non-involvement, delay in achieving the objectives. 

If you have made your own repository I presume that your work will focus there abandoning what exists here in OM. I don't consider it a timely decision because there are still issues and PR that need to be analyzed and integrated. We had barely managed to achieve the impossible by decreasing the number of issues and PR's and we are talking about projects whose finality is not known. For better or worse OM offers us an umbrella for everyone at this moment.

I would have preferred to discuss first any important changes you intend to make than to be made to choose whether to follow you or not. Honestly speaking, you have created some PR's lately that I would have liked to discuss before publishing. I also feel a certain upset in your message, a personal disappointment. You are an excellent programmer and organizer and for sure I would miss you.

**Reply by fballiano** [reply]
I don't know where you read all of this disappointment cause there's none.
I just don't like sitting and waiting when I have time to experiment, that's why I experiment and I hope somebody says "ah nice idea let's take it from there".

The PRs I opened and not written in stone so now it's the time to discuss them, in the PR itself. In some of them I specifically wrote "feedback and suggestions are welcome".

**Reply by addison74** [reply]
Nowadays it becomes more and more difficult to express a personal opinion because someone will come to tell you that it is wrong. I'm sorry if my previous message upset you in any way. I am not in the category of those who are waiting or sitting down, but I think that teamwork involves other more subtle things. I will think seriously about what is happening here and I will make a decision.

**Reply by fballiano** [reply]
I'm not upset in any way. This branch that I created is a test to show what's my vision. What I hope? That it will become OM v21, I dislike fragmentation and FLOSS should be about getting together.
But again, I've a bit of time lately and I'll try to produce stuff, I hope I'll be able to open more PRs soon since I've some ideas. And all my PRs (when I think they're something that should be integrated in the core) will be in OM, not anywhere else.

**Reply by fballiano** [reply]
Apologise if perhaps I write in a "too assertive" way and that seems I'm upset but I'm really not.

**Comment by elidrissidev** [comment]
I'd like to see a complete implementation of the REST API.

**Reply by fballiano** [reply]
I've never used the REST API so... what's missing in your opinion?

**Reply by elidrissidev** [reply]
A bunch of stuff, adding comments to orders, CRU* operations for invoices, shipments, coupons, etc...

**Reply by mageaustralia** [reply]
> A bunch of stuff, adding comments to orders, CRU* operations for invoices, shipments, coupons, etc...

I've just submitted a PR to Maho that has all off this and more.

**Reply by sreichel** [reply]
@mageaustralia hey, why not adding to OM?

**Reply by mageaustralia** [reply]
@sreichel  - I've been working with Mage1 -> OpenMage since 2008. Maho is what I always wanted Mage 2 / OpenMage to be - Doctrine (postgres / sqlite support)  / Symfony / VanillaJS  and more. For me personally, OpenMage became stale, barely progressed and I say that with a lot of respect to all the maintainers, really because you guys have done an amazing job continuing to support it but it really mostly felt like it hit the support stage and never moved on from that.

For reasons above,  we're migrating from OM -> Maho. For example: https://2630.mageaustralia.com.au/ - is using fly.io + neon.com (postgres) - basically offering *simple* autoscaling - It's beautiful and cheap as chips. 

When running my SSR  'headless' system on Cloudflare: https://demo.mageaustralia.com.au/ -  600rq/s is very achievable with low cost ($5 p/m with Cloudflare), easy to understand / clone / fork infrastructure. I've already [built a bunch of modules for Maho](https://github.com/mageaustralia?tab=repositories) - Gift Cards, Social Login, MeiliSearch, Stripe the API now and a POS system that's mostly complete. All released as open source.

I just want the best path forward for me and my clients and I feel this is it.

**Reply by sreichel** [reply]
Thanks. I understand.

Fab does very well with Maho. And yes, indeed, OM became lil stale last months.

imho ... the main difference is, OM always tried to keep BC promise. Fab is not that "limitated" an can push changes w/o taking care. 

Doctrine and  other Symfony-replacements are on its way. Only thing i am plan to port is frontend-theme w/o prototype.

> and never moved on from that.

Mhhh ... yes, OM is slower in progess. At the other hand ... Fab takes every PR from here and can spent his time for his things.  If i'd backport in same freqency he does OM and Maho would nearly look same. 

In longer term ... never really followed Maho, but tracked some changes ...

- when I started adding UnitTest, he added his pest tests (to stay different?)
- when I worked on n98-magerun, he started to add CLI commands (to stay different?)
- when new issues/PRs are opend, you`ll a commit there soon ....
- ...
- website built on mkdocs (Fab was against it?)
- ...
- replacing Zend ... whos idea? 
- ...

Thats, absolutly fine ...

## To speed up the work here ....

Take time for reviews!!!

Leave coments to PRs / issues.

**Be part of the community.**

**Reply by mageaustralia** [reply]
Hi @sreichel - you sound a little bitter or upset. I'm sorry about that. I mean open source is done with the understanding anyone can take what we do and run with it, for me that's the wonderful thing about it. I'm not really into politics, I'm just doing things that interest and excite me. I did hear from a 3rd party there was some dispute but I couldn't really find what it was about exactly and I'm not really that interested. 

I wish it was easy to backport the stuff I'm doing to OM - some of the modules probably could be, like the Stripe module: https://github.com/mageaustralia/maho-module-stripe

I simply don't have the bandwidth to do it myself but I'm more than happy if someone wants to take any of my work and bring it back into OpenMage - never any bad blood with anything I release Open Source.

If someone can get the API stuff back ported, I'd spend the extra little time getting my [headless frontend ](https://docs.mageaustralia.com.au/)working with it.

**Reply by sreichel** [reply]
No, im am not. I only worry about that some good ideas went to "next"/v21 branch only and have to be backported.

Another thing ... i am not sure about ... are people affraid to leave feedback? I am pretty sure we have a lot experienced developer around here. Every single review would help to speed up progess.

Id be glad to help with backporting ... let me know.

**Comment by kiatng** [comment]
I do not know how to use the Rest API, I tried but it didn't get anywhere. I used the old SOAP. Then I learned JSON-RPC from Colin, now it's the API that I use for projects. JSON-RPC is very lightweight and easy to implement. It is a direct replacement to SOAP in OM. The same endpoints for SOAP/XML-RPC work right of the box for JSON-RPC.

**Reply by elidrissidev** [reply]
Indeed the default authentication method for the REST API: OAuth 1, is very difficult to use and it's not possible to use for authenticating standalone scripts. I made my own extension to add support for HTTP Basic authentication which greatly simplified that.

My biggest concern with the SOAP API is its slow performance.

**Reply by fballiano** [reply]
https://github.com/OpenMage/magento-lts/pull/2273 :-)
